keras/metrics.py

```diff
@@ -20,6 +21,133 @@
 from .losses import poisson
 from .losses import cosine_proximity
 from .utils.generic_utils import deserialize_keras_object
+from .engine.topology import _to_snake_case
+
+
+class GlobalMetric(object):
+    """Base class for global metrics, which persist over epochs. Global Metrics
+    must inherit from this class.
+
+    # Properties:
+        __name__
+
+    # Methods:
+        update_states(self. y_true, y_pred)
+        reset_states(self)
+    """
+    def __init__(self):
+        """Initialize the Global Metric and give it a name.
+        """
+        # Instance name is class name with underscores
+        cls_name = self.__class__.__name__
+        self.__name__ = _to_snake_case(cls_name)
+
+    def __call__(self, y_true, y_pred):
+        return self.update_states(y_true, y_pred)
```

Comment:
I incorporated the ``self.add_update`` change.

Related to the second point, I think metrics being derived from the ``layer`` class would be rather confusing/unintuitive to people implementing their own stateful metrics. Even if metrics was a subclass of ``layers`` they would still require custom handling inside the training loop because of the existence of both stateful and non-stateful metrics in ``model.metrics``.